File: packages/sdk/src/utils/validation.ts
/**
 * ProtocolBanks SDK - Input Validation
 * 
 * 地址验证、金额验证、同形字符检测
 * Supports: Ethereum, Solana, Bitcoin addresses
 */

import type { ChainId, TokenSymbol, HomoglyphDetails } from '../types';
import { ErrorCodes } from '../types';
import { ProtocolBanksError } from './errors';
import { MAX_AMOUNT, MIN_AMOUNT, MAX_EXPIRY_HOURS, MAX_BATCH_SIZE } from '../config/chains';

// ============================================================================
// Homoglyph Detection
// ============================================================================

/**
 * Cyrillic characters that look like Latin characters
 * These are commonly used in homoglyph attacks
 */
const CYRILLIC_HOMOGLYPHS: Record<string, string> = {
  'а': 'a',  // Cyrillic а -> Latin a
  'е': 'e',  // Cyrillic е -> Latin e
  'о': 'o',  // Cyrillic о -> Latin o
  'р': 'p',  // Cyrillic р -> Latin p
  'с': 'c',  // Cyrillic с -> Latin c
  'х': 'x',  // Cyrillic х -> Latin x
  'у': 'y',  // Cyrillic у -> Latin y
  'А': 'A',  // Cyrillic А -> Latin A
  'В': 'B',  // Cyrillic В -> Latin B
  'Е': 'E',  // Cyrillic Е -> Latin E
  'К': 'K',  // Cyrillic К -> Latin K
  'М': 'M',  // Cyrillic М -> Latin M
  'Н': 'H',  // Cyrillic Н -> Latin H
  'О': 'O',  // Cyrillic О -> Latin O
  'Р': 'P',  // Cyrillic Р -> Latin P
  'С': 'C',  // Cyrillic С -> Latin C
  'Т': 'T',  // Cyrillic Т -> Latin T
  'Х': 'X',  // Cyrillic Х -> Latin X
};

/**
 * Greek characters that look like Latin characters
 */
const GREEK_HOMOGLYPHS: Record<string, string> = {
  'Α': 'A',  // Greek Alpha -> Latin A
  'Β': 'B',  // Greek Beta -> Latin B
  'Ε': 'E',  // Greek Epsilon -> Latin E
  'Ζ': 'Z',  // Greek Zeta -> Latin Z
  'Η': 'H',  // Greek Eta -> Latin H
  'Ι': 'I',  // Greek Iota -> Latin I
  'Κ': 'K',  // Greek Kappa -> Latin K
  'Μ': 'M',  // Greek Mu -> Latin M
  'Ν': 'N',  // Greek Nu -> Latin N
  'Ο': 'O',  // Greek Omicron -> Latin O
  'Ρ': 'P',  // Greek Rho -> Latin P
  'Τ': 'T',  // Greek Tau -> Latin T
  'Υ': 'Y',  // Greek Upsilon -> Latin Y
  'Χ': 'X',  // Greek Chi -> Latin X
  'ο': 'o',  // Greek omicron -> Latin o
};

/** All homoglyphs combined */
const ALL_HOMOGLYPHS: Record<string, string> = {
  ...CYRILLIC_HOMOGLYPHS,
  ...GREEK_HOMOGLYPHS,
};

/**
 * Detect homoglyph characters in a string
 */
export function detectHomoglyphs(input: string): HomoglyphDetails | null {
  const detectedCharacters: HomoglyphDetails['detectedCharacters'] = [];
  
  for (let i = 0; i < input.length; i++) {
    const char = input[i]!;
    const expected = ALL_HOMOGLYPHS[char];
    
    if (expected) {
      detectedCharacters.push({
        position: i,
        character: char,
        unicodePoint: `U+${char.charCodeAt(0).toString(16).toUpperCase().padStart(4, '0')}`,
        expectedCharacter: expected,
      });
    }
  }
  
  if (detectedCharacters.length === 0) {
    return null;
  }
  
  return {
    originalAddress: input,
    detectedCharacters,
  };
}

/**
 * Check if string contains homoglyphs
 */
export function containsHomoglyphs(input: string): boolean {
  return detectHomoglyphs(input) !== null;
}

/**
 * Normalize string by replacing homoglyphs with Latin equivalents
 */
export function normalizeHomoglyphs(input: string): string {
  let result = '';
  for (const char of input) {
    result += ALL_HOMOGLYPHS[char] ?? char;
  }
  return result;
}

// ============================================================================
// Address Validation
// ============================================================================

/** EVM address pattern (0x + 40 hex chars) */
const EVM_ADDRESS_PATTERN = /^0x[a-fA-F0-9]{40}$/;

/** Solana address pattern (Base58, 32-44 chars) */
const SOLANA_ADDRESS_PATTERN = /^[1-9A-HJ-NP-Za-km-z]{32,44}$/;

/** Bitcoin address patterns */
const BITCOIN_PATTERNS = {
  // Legacy (P2PKH) - starts with 1
  legacy: /^1[a-km-zA-HJ-NP-Z1-9]{25,34}$/,
  // SegWit (P2SH) - starts with 3
  segwit: /^3[a-km-zA-HJ-NP-Z1-9]{25,34}$/,
  // Native SegWit (Bech32) - starts with bc1
  bech32: /^bc1[a-z0-9]{39,59}$/,
  // Taproot (Bech32m) - starts with bc1p
  taproot: /^bc1p[a-z0-9]{58}$/,
};

/**
 * Validate EVM address (Ethereum, Polygon, etc.)
 */
export function isValidEVMAddress(address: string): boolean {
  // Check for homoglyphs first
  if (containsHomoglyphs(address)) {
    return false;
  }
  
  return EVM_ADDRESS_PATTERN.test(address);
}

/**
 * Validate Solana address
 */
export function isValidSolanaAddress(address: string): boolean {
  // Check for homoglyphs first
  if (containsHomoglyphs(address)) {
    return false;
  }
  
  return SOLANA_ADDRESS_PATTERN.test(address);
}

/**
 * Validate Bitcoin address
 */
export function isValidBitcoinAddress(address: string): boolean {
  // Check for homoglyphs first
  if (containsHomoglyphs(address)) {
    return false;
  }
  
  return (
    BITCOIN_PATTERNS.legacy.test(address) ||
    BITCOIN_PATTERNS.segwit.test(address) ||
    BITCOIN_PATTERNS.bech32.test(address) ||
    BITCOIN_PATTERNS.taproot.test(address)
  );
}

/**
 * Validate address for a specific chain
 */
export function isValidAddress(address: string, chainId?: ChainId): boolean {
  // Check for homoglyphs first
  if (containsHomoglyphs(address)) {
    return false;
  }
  
  if (chainId === 'solana') {
    return isValidSolanaAddress(address);
  }
  
  if (chainId === 'bitcoin') {
    return isValidBitcoinAddress(address);
  }
  
  // Default to EVM validation
  return isValidEVMAddress(address);
}

/**
 * Validate address and throw error if invalid
 */
export function validateAddress(address: string, chainId?: ChainId): void {
  // Check for homoglyphs
  const homoglyphDetails = detectHomoglyphs(address);
  if (homoglyphDetails) {
    throw new ProtocolBanksError({
      code: ErrorCodes.LINK_HOMOGLYPH_DETECTED,
      category: 'LINK',
      message: 'Potential homoglyph attack detected in address',
      details: homoglyphDetails,
      retryable: false,
    });
  }
  
  // Validate format
  if (!isValidAddress(address, chainId)) {
    const chainName = chainId === 'solana' ? 'Solana' : 
                      chainId === 'bitcoin' ? 'Bitcoin' : 'EVM';
    throw new ProtocolBanksError({
      code: ErrorCodes.LINK_INVALID_ADDRESS,
      category: 'LINK',
      message: `Invalid ${chainName} address format`,
      details: { address, chainId },
      retryable: false,
    });
  }
}

/**
 * Get address type
 */
export function getAddressType(address: string): 'evm' | 'solana' | 'bitcoin' | 'unknown' {
  if (isValidEVMAddress(address)) return 'evm';
  if (isValidSolanaAddress(address)) return 'solana';
  if (isValidBitcoinAddress(address)) return 'bitcoin';
  return 'unknown';
}

// ============================================================================
// Amount Validation
// ============================================================================

/**
 * Validate payment amount
 */
export function isValidAmount(amount: string): boolean {
  const num = parseFloat(amount);
  
  if (isNaN(num)) return false;
  if (num <= 0) return false;
  if (num < parseFloat(MIN_AMOUNT)) return false;
  if (num > parseFloat(MAX_AMOUNT)) return false;
  
  // Check for reasonable decimal places (max 18)
  const parts = amount.split('.');
  if (parts[1] && parts[1].length > 18) return false;
  
  return true;
}

/**
 * Validate amount and throw error if invalid
 */
export function validateAmount(amount: string): void {
  const num = parseFloat(amount);
  
  if (isNaN(num)) {
    throw new ProtocolBanksError({
      code: ErrorCodes.LINK_INVALID_AMOUNT,
      category: 'LINK',
      message: 'Amount must be a valid number',
      details: { amount },
      retryable: false,
    });
  }
  
  if (num <= 0) {
    throw new ProtocolBanksError({
      code: ErrorCodes.LINK_INVALID_AMOUNT,
      category: 'LINK',
      message: 'Amount must be greater than 0',
      details: { amount },
      retryable: false,
    });
  }
  
  if (num < parseFloat(MIN_AMOUNT)) {
    throw new ProtocolBanksError({
      code: ErrorCodes.LINK_INVALID_AMOUNT,
      category: 'LINK',
      message: `Amount must be at least ${MIN_AMOUNT}`,
      details: { amount, minAmount: MIN_AMOUNT },
      retryable: false,
    });
  }
  
  if (num > parseFloat(MAX_AMOUNT)) {
    throw new ProtocolBanksError({
      code: ErrorCodes.LINK_INVALID_AMOUNT,
      category: 'LINK',
      message: `Amount must not exceed ${MAX_AMOUNT}`,
      details: { amount, maxAmount: MAX_AMOUNT },
      retryable: false,
    });
  }
}

// ============================================================================
// Token Validation
// ============================================================================

/** Supported tokens */
const SUPPORTED_TOKENS: TokenSymbol[] = ['USDC', 'USDT', 'DAI', 'ETH', 'MATIC', 'BNB', 'SOL', 'BTC'];

/**
 * Validate token symbol
 */
export function isValidToken(token: string): token is TokenSymbol {
  return SUPPORTED_TOKENS.includes(token.toUpperCase() as TokenSymbol);
}

/**
 * Validate token and throw error if invalid
 */
export function validateToken(token: string): void {
  if (!isValidToken(token)) {
    throw new ProtocolBanksError({
      code: ErrorCodes.LINK_INVALID_TOKEN,
      category: 'LINK',
      message: `Unsupported token: ${token}`,
      details: { token, supportedTokens: SUPPORTED_TOKENS },
      retryable: false,
    });
  }
}

// ============================================================================
// Chain Validation
// ============================================================================

/** Supported chain IDs */
const SUPPORTED_CHAIN_IDS: ChainId[] = [1, 137, 8453, 42161, 10, 56, 'solana', 'bitcoin'];

/**
 * Validate chain ID
 */
export function isValidChainId(chainId: unknown): chainId is ChainId {
  return SUPPORTED_CHAIN_IDS.includes(chainId as ChainId);
}

/**
 * Validate chain and throw error if invalid
 */
export function validateChainId(chainId: unknown): void {
  if (!isValidChainId(chainId)) {
    throw new ProtocolBanksError({
      code: ErrorCodes.LINK_INVALID_CHAIN,
      category: 'LINK',
      message: `Unsupported chain: ${String(chainId)}`,
      details: { chainId, supportedChains: SUPPORTED_CHAIN_IDS },
      retryable: false,
    });
  }
}

// ============================================================================
// Expiry Validation
// ============================================================================

/**
 * Validate expiry hours
 */
export function isValidExpiryHours(hours: number): boolean {
  return Number.isInteger(hours) && hours >= 1 && hours <= MAX_EXPIRY_HOURS;
}

/**
 * Validate expiry and throw error if invalid
 */
export function validateExpiryHours(hours: number): void {
  if (!isValidExpiryHours(hours)) {
    throw new ProtocolBanksError({
      code: ErrorCodes.LINK_INVALID_EXPIRY,
      category: 'LINK',
      message: `Expiry hours must be between 1 and ${MAX_EXPIRY_HOURS}`,
      details: { hours, maxHours: MAX_EXPIRY_HOURS },
      retryable: false,
    });
  }
}

/**
 * Check if timestamp is expired
 */
export function isExpired(expiryTimestamp: number): boolean {
  return Date.now() > expiryTimestamp;
}

// ============================================================================
// Batch Validation
// ============================================================================

/**
 * Validate batch size
 */
export function isValidBatchSize(size: number): boolean {
  return Number.isInteger(size) && size >= 1 && size <= MAX_BATCH_SIZE;
}

/**
 * Validate batch size and throw error if invalid
 */
export function validateBatchSize(size: number): void {
  if (!isValidBatchSize(size)) {
    throw new ProtocolBanksError({
      code: ErrorCodes.BATCH_SIZE_EXCEEDED,
      category: 'BATCH',
      message: `Batch size must be between 1 and ${MAX_BATCH_SIZE}`,
      details: { size, maxSize: MAX_BATCH_SIZE },
      retryable: false,
    });
  }
}

// ============================================================================
// URL Validation
// ============================================================================

/**
 * Validate URL is HTTPS
 */
export function isHttpsUrl(url: string): boolean {
  try {
    const parsed = new URL(url);
    return parsed.protocol === 'https:';
  } catch {
    return false;
  }
}

/**
 * Validate URL and throw error if not HTTPS
 */
export function validateHttpsUrl(url: string, fieldName: string = 'URL'): void {
  if (!isHttpsUrl(url)) {
    throw new ProtocolBanksError({
      code: ErrorCodes.VALID_INVALID_FORMAT,
      category: 'VALID',
      message: `${fieldName} must be a valid HTTPS URL`,
      details: { url, fieldName },
      retryable: false,
    });
  }
}

// ============================================================================
// API Key Validation
// ============================================================================

/** API key patterns */
const API_KEY_PATTERN = /^pk_(live|test|sandbox)_[a-zA-Z0-9]{24,}$/;
const API_SECRET_PATTERN = /^sk_(live|test|sandbox)_[a-zA-Z0-9]{32,}$/;

/**
 * Validate API key format
 */
export function isValidApiKey(apiKey: string): boolean {
  return API_KEY_PATTERN.test(apiKey);
}

/**
 * Validate API secret format
 */
export function isValidApiSecret(apiSecret: string): boolean {
  return API_SECRET_PATTERN.test(apiSecret);
}

/**
 * Validate API credentials
 */
export function validateApiCredentials(apiKey: string, apiSecret: string): void {
  if (!isValidApiKey(apiKey)) {
    throw new ProtocolBanksError({
      code: ErrorCodes.AUTH_INVALID_API_KEY,
      category: 'AUTH',
      message: 'Invalid API key format',
      retryable: false,
    });
  }
  
  if (!isValidApiSecret(apiSecret)) {
    throw new ProtocolBanksError({
      code: ErrorCodes.AUTH_INVALID_SECRET,
      category: 'AUTH',
      message: 'Invalid API secret format',
      retryable: false,
    });
  }
}

// ============================================================================
// Memo Validation
// ============================================================================

/** Maximum memo length */
const MAX_MEMO_LENGTH = 256;

/**
 * Validate memo
 */
export function isValidMemo(memo: string): boolean {
  return memo.length <= MAX_MEMO_LENGTH;
}

/**
 * Validate memo and throw error if invalid
 */
export function validateMemo(memo: string): void {
  if (!isValidMemo(memo)) {
    throw new ProtocolBanksError({
      code: ErrorCodes.VALID_OUT_OF_RANGE,
      category: 'VALID',
      message: `Memo must not exceed ${MAX_MEMO_LENGTH} characters`,
      details: { length: memo.length, maxLength: MAX_MEMO_LENGTH },
      retryable: false,
    });
  }
}

File: packages/python/src/protocolbanks/utils/validation.py
"""
ProtocolBanks SDK - Input Validation

地址验证、金额验证、同形字符检测
Supports: Ethereum, Solana, Bitcoin addresses
"""

from __future__ import annotations

import re
import time
from typing import Any

from protocolbanks.config import MAX_AMOUNT, MAX_BATCH_SIZE, MAX_EXPIRY_HOURS, MIN_AMOUNT
from protocolbanks.errors import ProtocolBanksError
from protocolbanks.types import ChainId, ErrorCodes, HomoglyphDetails, TokenSymbol


# ============================================================================
# Homoglyph Detection
# ============================================================================

# Cyrillic characters that look like Latin characters
CYRILLIC_HOMOGLYPHS: dict[str, str] = {
    "а": "a",  # Cyrillic а -> Latin a
    "е": "e",  # Cyrillic е -> Latin e
    "о": "o",  # Cyrillic о -> Latin o
    "р": "p",  # Cyrillic р -> Latin p
    "с": "c",  # Cyrillic с -> Latin c
    "х": "x",  # Cyrillic х -> Latin x
    "у": "y",  # Cyrillic у -> Latin y
    "А": "A",  # Cyrillic А -> Latin A
    "В": "B",  # Cyrillic В -> Latin B
    "Е": "E",  # Cyrillic Е -> Latin E
    "К": "K",  # Cyrillic К -> Latin K
    "М": "M",  # Cyrillic М -> Latin M
    "Н": "H",  # Cyrillic Н -> Latin H
    "О": "O",  # Cyrillic О -> Latin O
    "Р": "P",  # Cyrillic Р -> Latin P
    "С": "C",  # Cyrillic С -> Latin C
    "Т": "T",  # Cyrillic Т -> Latin T
    "Х": "X",  # Cyrillic Х -> Latin X
}

# Greek characters that look like Latin characters
GREEK_HOMOGLYPHS: dict[str, str] = {
    "Α": "A",  # Greek Alpha -> Latin A
    "Β": "B",  # Greek Beta -> Latin B
    "Ε": "E",  # Greek Epsilon -> Latin E
    "Ζ": "Z",  # Greek Zeta -> Latin Z
    "Η": "H",  # Greek Eta -> Latin H
    "Ι": "I",  # Greek Iota -> Latin I
    "Κ": "K",  # Greek Kappa -> Latin K
    "Μ": "M",  # Greek Mu -> Latin M
    "Ν": "N",  # Greek Nu -> Latin N
    "Ο": "O",  # Greek Omicron -> Latin O
    "Ρ": "P",  # Greek Rho -> Latin P
    "Τ": "T",  # Greek Tau -> Latin T
    "Υ": "Y",  # Greek Upsilon -> Latin Y
    "Χ": "X",  # Greek Chi -> Latin X
    "ο": "o",  # Greek omicron -> Latin o
}

# All homoglyphs combined
ALL_HOMOGLYPHS: dict[str, str] = {**CYRILLIC_HOMOGLYPHS, **GREEK_HOMOGLYPHS}


def detect_homoglyphs(input_str: str) -> HomoglyphDetails | None:
    """Detect homoglyph characters in a string"""
    detected_characters: list[dict[str, Any]] = []

    for i, char in enumerate(input_str):
        expected = ALL_HOMOGLYPHS.get(char)
        if expected:
            detected_characters.append(
                {
                    "position": i,
                    "character": char,
                    "unicode_point": f"U+{ord(char):04X}",
                    "expected_character": expected,
                }
            )

    if not detected_characters:
        return None

    return HomoglyphDetails(
        original_address=input_str,
        detected_characters=detected_characters,
    )


def contains_homoglyphs(input_str: str) -> bool:
    """Check if string contains homoglyphs"""
    return detect_homoglyphs(input_str) is not None


def normalize_homoglyphs(input_str: str) -> str:
    """Normalize string by replacing homoglyphs with Latin equivalents"""
    result = ""
    for char in input_str:
        result += ALL_HOMOGLYPHS.get(char, char)
    return result


# ============================================================================
# Address Validation
# ============================================================================

# EVM address pattern (0x + 40 hex chars)
EVM_ADDRESS_PATTERN = re.compile(r"^0x[a-fA-F0-9]{40}$")

# Solana address pattern (Base58, 32-44 chars)
SOLANA_ADDRESS_PATTERN = re.compile(r"^[1-9A-HJ-NP-Za-km-z]{32,44}$")

# Bitcoin address patterns
BITCOIN_PATTERNS = {
    "legacy": re.compile(r"^1[a-km-zA-HJ-NP-Z1-9]{25,34}$"),  # P2PKH
    "segwit": re.compile(r"^3[a-km-zA-HJ-NP-Z1-9]{25,34}$"),  # P2SH
    "bech32": re.compile(r"^bc1[a-z0-9]{39,59}$"),  # Native SegWit
    "taproot": re.compile(r"^bc1p[a-z0-9]{58}$"),  # Taproot
}


def is_valid_evm_address(address: str) -> bool:
    """Validate EVM address (Ethereum, Polygon, etc.)"""
    if contains_homoglyphs(address):
        return False
    return bool(EVM_ADDRESS_PATTERN.match(address))


def is_valid_solana_address(address: str) -> bool:
    """Validate Solana address"""
    if contains_homoglyphs(address):
        return False
    return bool(SOLANA_ADDRESS_PATTERN.match(address))


def is_valid_bitcoin_address(address: str) -> bool:
    """Validate Bitcoin address"""
    if contains_homoglyphs(address):
        return False
    return any(pattern.match(address) for pattern in BITCOIN_PATTERNS.values())


def is_valid_address(address: str, chain_id: ChainId | None = None) -> bool:
    """Validate address for a specific chain"""
    if contains_homoglyphs(address):
        return False

    if chain_id == "solana":
        return is_valid_solana_address(address)

    if chain_id == "bitcoin":
        return is_valid_bitcoin_address(address)

    # Default to EVM validation
    return is_valid_evm_address(address)


def validate_address(address: str, chain_id: ChainId | None = None) -> None:
    """Validate address and throw error if invalid"""
    # Check for homoglyphs
    homoglyph_details = detect_homoglyphs(address)
    if homoglyph_details:
        raise ProtocolBanksError(
            code=ErrorCodes.LINK_HOMOGLYPH_DETECTED,
            message="Potential homoglyph attack detected in address",
            details=homoglyph_details.model_dump(),
            retryable=False,
        )

    # Validate format
    if not is_valid_address(address, chain_id):
        chain_name = (
            "Solana" if chain_id == "solana" else "Bitcoin" if chain_id == "bitcoin" else "EVM"
        )
        raise ProtocolBanksError(
            code=ErrorCodes.LINK_INVALID_ADDRESS,
            message=f"Invalid {chain_name} address format",
            details={"address": address, "chain_id": chain_id},
            retryable=False,
        )


def get_address_type(address: str) -> str:
    """Get address type"""
    if is_valid_evm_address(address):
        return "evm"
    if is_valid_solana_address(address):
        return "solana"
    if is_valid_bitcoin_address(address):
        return "bitcoin"
    return "unknown"


# ============================================================================
# Amount Validation
# ============================================================================


def is_valid_amount(amount: str) -> bool:
    """Validate payment amount"""
    try:
        num = float(amount)
        if num <= 0:
            return False
        if num < float(MIN_AMOUNT):
            return False
        if num > float(MAX_AMOUNT):
            return False

        # Check for reasonable decimal places (max 18)
        parts = amount.split(".")
        if len(parts) > 1 and len(parts[1]) > 18:
            return False

        return True
    except (ValueError, TypeError):
        return False


def validate_amount(amount: str) -> None:
    """Validate amount and throw error if invalid"""
    try:
        num = float(amount)
    except (ValueError, TypeError):
        raise ProtocolBanksError(
            code=ErrorCodes.LINK_INVALID_AMOUNT,
            message="Amount must be a valid number",
            details={"amount": amount},
            retryable=False,
        )

    if num <= 0:
        raise ProtocolBanksError(
            code=ErrorCodes.LINK_INVALID_AMOUNT,
            message="Amount must be greater than 0",
            details={"amount": amount},
            retryable=False,
        )

    if num < float(MIN_AMOUNT):
        raise ProtocolBanksError(
            code=ErrorCodes.LINK_INVALID_AMOUNT,
            message=f"Amount must be at least {MIN_AMOUNT}",
            details={"amount": amount, "min_amount": MIN_AMOUNT},
            retryable=False,
        )

    if num > float(MAX_AMOUNT):
        raise ProtocolBanksError(
            code=ErrorCodes.LINK_INVALID_AMOUNT,
            message=f"Amount must not exceed {MAX_AMOUNT}",
            details={"amount": amount, "max_amount": MAX_AMOUNT},
            retryable=False,
        )


# ============================================================================
# Token Validation
# ============================================================================

SUPPORTED_TOKENS: list[TokenSymbol] = ["USDC", "USDT", "DAI", "ETH", "MATIC", "BNB", "SOL", "BTC"]


def is_valid_token(token: str) -> bool:
    """Validate token symbol"""
    return token.upper() in SUPPORTED_TOKENS


def validate_token(token: str) -> None:
    """Validate token and throw error if invalid"""
    if not is_valid_token(token):
        raise ProtocolBanksError(
            code=ErrorCodes.LINK_INVALID_TOKEN,
            message=f"Unsupported token: {token}",
            details={"token": token, "supported_tokens": SUPPORTED_TOKENS},
            retryable=False,
        )


# ============================================================================
# Chain Validation
# ============================================================================

SUPPORTED_CHAIN_IDS: list[ChainId] = [1, 137, 8453, 42161, 10, 56, "solana", "bitcoin"]


def is_valid_chain_id(chain_id: Any) -> bool:
    """Validate chain ID"""
    return chain_id in SUPPORTED_CHAIN_IDS


def validate_chain_id(chain_id: Any) -> None:
    """Validate chain and throw error if invalid"""
    if not is_valid_chain_id(chain_id):
        raise ProtocolBanksError(
            code=ErrorCodes.LINK_INVALID_CHAIN,
            message=f"Unsupported chain: {chain_id}",
            details={"chain_id": chain_id, "supported_chains": SUPPORTED_CHAIN_IDS},
            retryable=False,
        )


# ============================================================================
# Expiry Validation
# ============================================================================


def is_valid_expiry_hours(hours: int) -> bool:
    """Validate expiry hours"""
    return isinstance(hours, int) and 1 <= hours <= MAX_EXPIRY_HOURS


def validate_expiry_hours(hours: int) -> None:
    """Validate expiry and throw error if invalid"""
    if not is_valid_expiry_hours(hours):
        raise ProtocolBanksError(
            code=ErrorCodes.LINK_INVALID_EXPIRY,
            message=f"Expiry hours must be between 1 and {MAX_EXPIRY_HOURS}",
            details={"hours": hours, "max_hours": MAX_EXPIRY_HOURS},
            retryable=False,
        )


def is_expired(expiry_timestamp: int) -> bool:
    """Check if timestamp is expired"""
    return int(time.time() * 1000) > expiry_timestamp


# ============================================================================
# Batch Validation
# ============================================================================


def is_valid_batch_size(size: int) -> bool:
    """Validate batch size"""
    return isinstance(size, int) and 1 <= size <= MAX_BATCH_SIZE


def validate_batch_size(size: int) -> None:
    """Validate batch size and throw error if invalid"""
    if not is_valid_batch_size(size):
        raise ProtocolBanksError(
            code=ErrorCodes.BATCH_SIZE_EXCEEDED,
            message=f"Batch size must be between 1 and {MAX_BATCH_SIZE}",
            details={"size": size, "max_size": MAX_BATCH_SIZE},
            retryable=False,
        )


# ============================================================================
# URL Validation
# ============================================================================


def is_https_url(url: str) -> bool:
    """Validate URL is HTTPS"""
    try:
        from urllib.parse import urlparse

        parsed = urlparse(url)
        return parsed.scheme == "https"
    except Exception:
        return False


def validate_https_url(url: str, field_name: str = "URL") -> None:
    """Validate URL and throw error if not HTTPS"""
    if not is_https_url(url):
        raise ProtocolBanksError(
            code=ErrorCodes.VALID_INVALID_FORMAT,
            message=f"{field_name} must be a valid HTTPS URL",
            details={"url": url, "field_name": field_name},
            retryable=False,
        )


# ============================================================================
# API Key Validation
# ============================================================================

API_KEY_PATTERN = re.compile(r"^pk_(live|test|sandbox)_[a-zA-Z0-9]{24,}$")
API_SECRET_PATTERN = re.compile(r"^sk_(live|test|sandbox)_[a-zA-Z0-9]{32,}$")


def is_valid_api_key(api_key: str) -> bool:
    """Validate API key format"""
    return bool(API_KEY_PATTERN.match(api_key))


def is_valid_api_secret(api_secret: str) -> bool:
    """Validate API secret format"""
    return bool(API_SECRET_PATTERN.match(api_secret))


def validate_api_credentials(api_key: str, api_secret: str) -> None:
    """Validate API credentials"""
    if not is_valid_api_key(api_key):
        raise ProtocolBanksError(
            code=ErrorCodes.AUTH_INVALID_API_KEY,
            message="Invalid API key format",
            retryable=False,
        )

    if not is_valid_api_secret(api_secret):
        raise ProtocolBanksError(
            code=ErrorCodes.AUTH_INVALID_SECRET,
            message="Invalid API secret format",
            retryable=False,
        )


# ============================================================================
# Memo Validation
# ============================================================================

MAX_MEMO_LENGTH = 256


def is_valid_memo(memo: str) -> bool:
    """Validate memo"""
    return len(memo) <= MAX_MEMO_LENGTH


def validate_memo(memo: str) -> None:
    """Validate memo and throw error if invalid"""
    if not is_valid_memo(memo):
        raise ProtocolBanksError(
            code=ErrorCodes.VALID_OUT_OF_RANGE,
            message=f"Memo must not exceed {MAX_MEMO_LENGTH} characters",
            details={"length": len(memo), "max_length": MAX_MEMO_LENGTH},
            retryable=False,
        )
